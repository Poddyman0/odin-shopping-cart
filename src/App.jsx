
import { useState, useEffect } from 'react'
import { useParams } from "react-router-dom";
import HomePage from './components/homepage'
import ShoppingCart from './components/shoppingcart'
import ShopPage from './components/shoppage'
import NavBar from './components/Navbar';
import WelcomePage from "./components/welcomepage"

function App() {
  const {name} = useParams()
  const [cartCount, setCartCount] = useState(0)
  const [cartItems, setCartItems] = useState([])
  const [shopPoductsArray, setShopPoductsArray] = useState([])
  useEffect(() => {
      fetch('https://fakestoreapi.com/products')
        .then(function(response) {
          return response.json()
        })
        .then(function(response) {  
          setShopPoductsArray(response)
        })
        .catch(function(err) {
            console.log("Error: ", err)
        });
    
  }, []);
  return (
  <>
    <NavBar  cartCount={cartCount} setCartCount={setCartCount} cartItems={cartItems} setCartItems={setCartItems} shopPoductsArray={shopPoductsArray}/>
    {name === "homepage" ? (
    <HomePage cartCount={cartCount} setCartCount={setCartCount} cartItems={cartItems} setCartItems={setCartItems} shopPoductsArray={shopPoductsArray}/>
    ) : name === "shopcart" ? (
      <ShoppingCart cartCount={cartCount} setCartCount={setCartCount} cartItems={cartItems} setCartItems={setCartItems} shopPoductsArray={shopPoductsArray}/>
    ) : name === "shoppage" ? (
      <ShopPage cartCount={cartCount} setCartCount={setCartCount} cartItems={cartItems} setCartItems={setCartItems} shopPoductsArray={shopPoductsArray}/>
    ) : (
      <WelcomePage />
    )}
  </>
  )
}

export default App

// test it using react testing library
// host it.
