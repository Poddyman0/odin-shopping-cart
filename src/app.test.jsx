import { describe, expect } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';
import App from './App';
import NavBar from './components/Navbar'
import ShopCard from './components/shopcard'
import ShopPage from './components/shoppage';
import ShoppingCart from './components/shoppingcart'
import WelcomePage from './components/welcomepage'
import HomePage from './components/homepage'
import CartItem from './components/cartitem'
import ControlledCarousel from './components/carousel'

describe('NavBar Component', () => {
    test('Navbar links point to correct destinations', () => {
      render(
        <BrowserRouter>
          <NavBar cartCount={3} />
        </BrowserRouter>
      );
  
      const shopPageLink = screen.getByText('Shop Page');
      expect(shopPageLink).toHaveAttribute('href', '/shoppage');
  
      const homepageLink = screen.getByText('Homepage');
      expect(homepageLink).toHaveAttribute('href', '/homepage');
  
      const shopLink = screen.getByText('Shop');
      expect(shopLink).toHaveAttribute('href', '/shoppage');
  
      const cartLink = screen.getByText('Cart');
      expect(cartLink).toHaveAttribute('href', '/shopcart');
    });
  
    test('Cart count is displayed correctly', () => {
      render(
        <BrowserRouter>
          <NavBar cartCount={5} />
        </BrowserRouter>
      );
  
      const cartAmount = screen.getByText('5 items');
      expect(cartAmount).toBeInTheDocument();
    });
})


  describe('WelcomePage Component', () => {
    test('welcome page content is displayed', () => {
        render(
            <BrowserRouter>
              <WelcomePage />
            </BrowserRouter>
          );

          const welcomeMessage = screen.getByTestId("welcome-message-test");
          expect(welcomeMessage).toHaveTextContent('Welcome to our shop!');
    })
  })

  describe('ShoppingCart and CartItem Component', () => {
    const cartItems = [
      { id: 1, image: 'pic1', title: 'Item 1', price: 10, amount: 1 },
      { id: 2, image: 'pic2', title: 'Item 2', price: 15, amount: 2 },
    ];
  
    test('Total cart bill amount is calculated correctly', () => {
      render(
        <BrowserRouter>
            <ShoppingCart cartItems={cartItems} />
        </BrowserRouter>);
        const totalAmountNum = screen.getByTestId('cart-total-num')
        expect(totalAmountNum).toHaveTextContent('£40')
    });


    test('CartItem component is rendered for each item in the ShoppingCart and content of each cartitem is rendered correctly', () => {
      render(        
        <BrowserRouter>
            <ShoppingCart cartItems={cartItems} />
        </BrowserRouter>);
      const cartItemElements = screen.getAllByTestId('cart-item');
      expect(cartItemElements.length).toBe(cartItems.length);


      const cartItemElementsTitles = screen.getAllByTestId('cart-item-test-title');
      expect(cartItemElementsTitles[0]).toHaveTextContent("Item 1")
      const cartItemElementsPic = screen.getAllByTestId('cart-item-test-image')
      expect(cartItemElementsPic[0]).toHaveAttribute('src', 'pic1')
      const cartItemElementsPrice = screen.getAllByTestId('cart-item-test-price');
      expect(cartItemElementsPrice[0]).toHaveTextContent("Price: £10")
      const cartItemElementsAmount = screen.getAllByTestId('cart-item-test-amount');
      expect(cartItemElementsAmount[0]).toHaveTextContent("Amount: 1")


    });
  
    test('Pay button is rendered', () => {
      render(<BrowserRouter>
        <ShoppingCart cartItems={cartItems} />
    </BrowserRouter>);
      const payButton = screen.getByRole('button', { name: 'Pay' });
      expect(payButton).toBeInTheDocument();
    });
});


describe('ShopCard and ShopPage Components', () => {
    test('content of each ShopCard item in ShopPage is rendered correctly', () => {
        const shopPoductsArray = [
            { id: 1, image: 'pic1', title: 'Item 1', price: 10, description: 'item 1 description', category: 'item 1 category', rating: {rate: 4.5, count: 25}},
            { id: 2, image: 'pic2', title: 'Item 2', price: 15, description: 'item 2 description', category: 'item 2 category', rating: {rate: 3.5, count: 36}},
          ];
        
        render(        
            <BrowserRouter>
                <ShopPage shopPoductsArray={shopPoductsArray}/>
            </BrowserRouter>
            );
        

        const shopItemElements = screen.getAllByTestId('shop-item');
        expect(shopItemElements.length).toBe(shopPoductsArray.length);

        const shopItemElementsPicture = screen.getAllByTestId('shop-item-picture')
        expect(shopItemElementsPicture[0]).toHaveAttribute('src', 'pic1')
        expect(shopItemElementsPicture[1]).toHaveAttribute('src', 'pic2')

        const shopCardTitle = screen.getAllByTestId('shop-item-title')
        expect(shopCardTitle[0]).toHaveTextContent("Item 1")
        expect(shopCardTitle[1]).toHaveTextContent("Item 2")


        const shopCardPrice = screen.getAllByTestId('shop-item-price')
        expect(shopCardPrice[0]).toHaveTextContent("Price: £10")
        expect(shopCardPrice[1]).toHaveTextContent("Price: £15")


        const shopCardDescription = screen.getAllByTestId('shop-item-description')
        expect(shopCardDescription[0]).toHaveTextContent("Description: item 1 description")
        expect(shopCardDescription[1]).toHaveTextContent("Description: item 2 description")


        const shopCardCategory = screen.getAllByTestId('shop-item-category')
        expect(shopCardCategory[0]).toHaveTextContent("Category: item 1 category")
        expect(shopCardCategory[1]).toHaveTextContent("Category: item 2 category")


        const shopCardRatingCount = screen.getAllByTestId('shop-item-rating-count')
        expect(shopCardRatingCount[0]).toHaveTextContent("25")
        expect(shopCardRatingCount[1]).toHaveTextContent("36")

        const shopCardAmountLabel = screen.getAllByTestId('shop-item-amount-label')
        expect(shopCardAmountLabel[0]).toHaveTextContent("Amount:")
        expect(shopCardAmountLabel[1]).toHaveTextContent("Amount:")

        const shopCardIncrementButton = screen.getAllByTestId('shop-item-increment-button')
        expect(shopCardIncrementButton[0]).toHaveTextContent("Increment Amount")
        expect(shopCardIncrementButton[1]).toHaveTextContent("Increment Amount")

        const shopCardDecrementButton = screen.getAllByTestId('shop-item-decrement-button')
        expect(shopCardDecrementButton[0]).toHaveTextContent("Decrement Amount")
        expect(shopCardDecrementButton[1]).toHaveTextContent("Decrement Amount")

        const shopCardAddToCartButton = screen.getAllByTestId('shop-item-add-to-cart-button')
        expect(shopCardAddToCartButton[0]).toHaveTextContent("Add To Cart")
        expect(shopCardAddToCartButton[1]).toHaveTextContent("Add To Cart")
        "shop-item-amount-input"
        
    })
    const shopPoductsArray = [
        { id: 1, image: 'pic1', title: 'Item 1', price: 10, description: 'item 1 description', category: 'item 1 category', rating: {rate: 4.5, count: 25}},
        { id: 2, image: 'pic2', title: 'Item 2', price: 15, description: 'item 2 description', category: 'item 2 category', rating: {rate: 3.5, count: 36}},
      ];
    

    const setup = () => {
        const utils = render(       
            <BrowserRouter>
                <ShopPage shopPoductsArray={shopPoductsArray}/>
            </BrowserRouter>
            )
        const allInputs = screen.getAllByTestId('shop-item-amount-input')
        const input = allInputs[0]
        const allIncrementButtons = screen.getAllByTestId('shop-item-increment-button')
        const incrementButton = allIncrementButtons[0]
        const allDecrementButtons = screen.getAllByTestId('shop-item-decrement-button')
        const decrementButton = allDecrementButtons[0]
        return {
            input,
            incrementButton,
            decrementButton,
            ...utils,
        }
    }
    test('ShopCard amount changes when typed in', () => {
        const {input} = setup()
        fireEvent.change(input, {target: {value: '67'}})
        expect(input.value).toBe('67')        
    })
    test('ShopCard amount input increases when increment button is clicked', () => {
        const {input, incrementButton} = setup()
        fireEvent.click(incrementButton)
        expect(input.value).toBe('1')        
    })
    test('ShopCard amount input decreases when increment button is clicked', () => {
        const {input, decrementButton} = setup()
        fireEvent.change(input, {target: {value: '67'}})
        fireEvent.click(decrementButton)
        expect(input.value).toBe('66')        
    })
})



describe('HomePage and Carousel Components', () => {
    test('renders homepage heading', () => {
        render(
        <BrowserRouter>
            <HomePage />
        </BrowserRouter>
        );
        const homepageElement = screen.getByTestId('homepage-welcome-message')
        expect(homepageElement).toHaveTextContent('Home Page')
    })
    test('renders homepage carousel pictures and text', () => {
        render(
        <BrowserRouter>
            <HomePage />
        </BrowserRouter>
        );
        const homepageCarouselCard = screen.getAllByTestId('carousel-card')
        expect(homepageCarouselCard.length).toBe(3);
        const homepageCarouselCardPicture = screen.getAllByTestId('carousel-card-picture')
        expect(homepageCarouselCardPicture[0]).toHaveAttribute('src', '/src/components/immages/picone.webp')
        const homepageCarouselCardTitle = screen.getAllByTestId('carousel-title')
        expect(homepageCarouselCardTitle[0]).toHaveTextContent("Ear Rings")
        const homepageCarouselCardText = screen.getAllByTestId('carousel-text')
        expect(homepageCarouselCardText[0]).toHaveTextContent("Nulla vitae elit libero, a pharetra augue mollis interdum.")

    })
})  




/*



- shopcart title

*/