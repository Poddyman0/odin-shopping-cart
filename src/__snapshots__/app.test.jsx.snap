// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShopCard Component > ShopCard 1`] = `<div />`;

exports[`ShopCard Component > renders ShopCard 1`] = `
<div>
  <nav
    class="navbar navbar-dark bg-dark fixed-top"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/shoppage"
      >
        Shop Page
      </a>
      <button
        aria-controls="offcanvasDarkNavbar"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-bs-target="#offcanvasDarkNavbar"
        data-bs-toggle="offcanvas"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        aria-labelledby="offcanvasDarkNavbarLabel"
        class="offcanvas offcanvas-end text-bg-dark"
        id="offcanvasDarkNavbar"
        tabindex="-1"
      >
        <div
          class="offcanvas-header"
        >
          <h5
            class="offcanvas-title"
            id="offcanvasDarkNavbarLabel"
          >
            Browse Shop
          </h5>
          <button
            aria-label="Close"
            class="btn-close btn-close-white"
            data-bs-dismiss="offcanvas"
            type="button"
          />
        </div>
        <div
          class="offcanvas-body"
        >
          <ul
            class="navbar-nav justify-content-end flex-grow-1 pe-3"
          >
            <li
              class="nav-item"
            >
              <a
                aria-current="page"
                class="nav-link"
                href="/homepage"
              >
                Homepage
                <svg
                  class="bi bi-house"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.707 1.5a1 1 0 0 0-1.414 0L.646 8.146a.5.5 0 0 0 .708.708L2 8.207V13.5A1.5 1.5 0 0 0 3.5 15h9a1.5 1.5 0 0 0 1.5-1.5V8.207l.646.647a.5.5 0 0 0 .708-.708L13 5.793V2.5a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5v1.293zM13 7.207V13.5a.5.5 0 0 1-.5.5h-9a.5.5 0 0 1-.5-.5V7.207l5-5z"
                  />
                </svg>
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                aria-current="page"
                class="nav-link"
                href="/shoppage"
              >
                Shop
                <svg
                  class="bi bi-bag"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 1a2.5 2.5 0 0 1 2.5 2.5V4h-5v-.5A2.5 2.5 0 0 1 8 1m3.5 3v-.5a3.5 3.5 0 1 0-7 0V4H1v10a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V4zM2 5h12v9a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1z"
                  />
                </svg>
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                aria-current="page"
                class="nav-link"
                href="/shopcart"
              >
                Cart
                <svg
                  class="bi bi-cart"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .491.592l-1.5 8A.5.5 0 0 1 13 12H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5M3.102 4l1.313 7h8.17l1.313-7zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4m7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4m-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2m7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2"
                  />
                </svg>
              </a>
              <a
                aria-current="page"
                class="nav-link"
                href="/shopcart"
              >
                <div
                  id="cart-amount-nav-bar"
                >
                   items
                </div>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </nav>
  <div
    id="shop-style"
  >
    <h1>
      Shop
    </h1>
    <div
      id="shop-card-container"
    />
  </div>
</div>
`;
